### PR TITLE
Add simple implementation of gqap & gwap for text paragraphs

### PR DESCRIPTION
**Pull request by gaganis**
I often edit markdown files and when on vim I use gqap to format their
paragraphs. I used to have to switch between IDE and vim just to
reformat my paragraphs. This commit introduces a simple, different from
vim & non-customizable version of gqap intented to satisfy reformatting
simple paragraphs for text files.

A key difference on how text is reformatted is that this version is not
preserving whitespace and replaces all whitespace inside the paragraph
with spaces.

Also vims reformatting in general is highly customizable with multiple
options eg textwidth, formatoptions, autoindent.

See also: VIM-1158

### CODE REVIEW

**Comment by gaganis** [comment]
Hi, 

I have created this implementation of gqap for my own use. I use gqap a lot when working on markdown files, and I was really missing this feature. I would work on these files using vim but if I forgot and used the ide I would then have to open vim do the gqap and then return again... tiresome.

This implementation works for my use case, I really do not need all the other features and I also do not need the whitespace preservation that vim does so I intend to use this.

I am planning to continue working on this to add a way to configure/customize the paragraph width like vim uses `textwidth`. I would like to explore this to the ideavimrc as textwith.

I struggled whether to post this PR due to it's limitations and differences from vim but I eventually decided positively because I find gqap to be very usefull and even this implementation can be useful to other people.

**Comment by david-drinn** [comment]
Is there a developer getting started guide for plugins or the IdeaVim plugin? I'm struggling to get up and running far enough to test this.

**Comment by AlexPl292** [comment]
@david-drinn Hi! The developer's guide can be found here: https://github.com/JetBrains/ideavim/blob/master/CONTRIBUTING.md

**Review by AlexPl292 — COMMENTED** [review]
Hi, thank you for your work! However, I'm a bit confused, you say that the main goal of this code is to support wrapping in md files. How does this formatting differ from the default IJ formatting for markdown files?

**Comment by AlexPl292** [comment]
Omg, wait. It looks like IJ supported md formatting since 2020.3 or something like that. Does md formatting work for you?

File: src/com/maddyhome/idea/vim/action/change/change/ReformatCodeMotionAction.kt [hunk 1]
```diff
@@ -40,6 +45,132 @@ class ReformatCodeMotionAction : ChangeEditorActionHandler.ForEachCaret(), Dupli
...
+    fun reformatParagraph(editor: Editor, caret: Caret, range: TextRange): Int {
+      val startOffset = range.startOffset
+      val endOffset = range.endOffset
+      val textRange = com.intellij.openapi.util.TextRange.create(startOffset, endOffset)
+      val text = editor.document.getText(textRange)
+      val oldCaretEditorOffset = caret.offset
+      val oldCaretParagraphOffset = oldCaretEditorOffset - startOffset
+
+      val formatResult: ReformatParagraphResult = reformatTextAsParagraph(text, oldCaretParagraphOffset)
+      editor.document.replaceString(startOffset, endOffset, formatResult.text)
```
[the remark this answers is not in the document]
Comment (gaganis) [reply src/com/maddyhome/idea/vim/action/change/change/ReformatCodeMotionAction.kt:88]: Are you specifically referring to  `editor.document.replaceString()`? 

Yes it is working.

**Comment by gaganis** [comment]
> Omg, wait. It looks like IJ supported md formatting since 2020.3 or something like that. Does md formatting work for you?

Heh we have been a bit late :disappointed:. I am happy  though that overall the intellij suite has progressed and it is covering this case :smiley:.

`gqap` sequence is now formatting paragraphs in `.md` files n 2020.3 with the released ideavim plugin. Without this PR we are sending the formatting command vanilla to intellij's formatter. So it worked _out of the box_ for the plugin.

***

Have to say though, to my credit :grinning: , that my formatting algorithm is handling text a bit better on some edge cases that I am aware of .  The following paragraph for example is not formatted by IJ as I would expect it to.

The original paragraph text:
```For my OS terminal(MATE Terminal running on Linux Mint 17.2) I found the
following from the powerline installation
documentation: [Installation on Linux] (https://powerline.readthedocs.io/en/latest/installation/linux.html#font-installation) where I following the instruction in the font-installation section which worked quite nicelly.
```

What intellij default md formatting does:
```
For my OS terminal(MATE Terminal running on Linux Mint 17.2) I found the
following from the powerline installation
documentation: [Installation on Linux] (https://powerline.readthedocs.io/en/latest/installation/linux.html#font-installation)
where I following the instruction in the font-installation section which worked
quite nicelly.
```

What I would expect and is produced by the algorithm I wrote for this PR: 
```
For my OS terminal(MATE Terminal running on Linux Mint 17.2) I found the
following from the powerline installation documentation: [Installation on Linux]
(https://powerline.readthedocs.io/en/latest/installation/linux.html#font-installation)
where I following the instruction in the font-installation section which worked
quite nicelly.
```

**Comment by gaganis** [comment]
@AlexPl292 can you get in me in touch with the guys that are writing the .md formatting code? Maybe my code or I could be helpful for them.

**Comment by AlexPl292** [comment]
Yeah, I see, your version really looks nicer :)
You can check how it works for markdown plugin, but I believe the plugin formats md of a different layer of abstraction. While you process the file as a text, md plugin processes the file as an AST. Anyway, I think this can be improved as well.

Okay, md plugin is located inside of IJ-community repository: https://github.com/JetBrains/intellij-community/tree/master/plugins/markdown
So, in case of PRs, just send them to `https://github.com/JetBrains/intellij-community`.
Not sure who works on this, but I believe this can be retrieved from the git history/blame.
The classes related to formatting can be found in https://github.com/JetBrains/intellij-community/blob/master/plugins/markdown/resource/META-INF/plugin.xml. Just search for "format" in this file.

Hope this would help. Good luck with that :D

**Pull request closed without merging**